src/matrix.h

```diff
@@ -20,96 +21,465 @@
-// (expands or contracts vectors differently in different axis)
-// (warning: this is not reversed by Invert_Simple)
-// (warning: do not apply this kind of matrix to direction vectors)
-void Matrix4x4_CreateScale3 (mat4_t *out, double x, double y, double z);
-// creates a matrix for a quake entity
-void Matrix4x4_CreateFromQuakeEntity(mat4_t *out, double x, double y, double z, double pitch, double yaw,
-                                     double roll, double scale);
-
-void Matrix4x4_CreateFromEntity(mat4_t *out, const vec3_t origin, const vec3_t angles, float scale);
-
-// creates a matrix4x4 from a float[16] array in the OpenGL orientation
-void Matrix4x4_FromArrayFloatGL(mat4_t *out, const float in[16]);
-
-// converts a matrix4x4 to a set of 3D vectors for the 3 axial directions, and the translate
-void Matrix4x4_ToVectors(const mat4_t *in, float vx[3], float vy[3], float vz[3], float t[3]);
-// creates a matrix4x4 from a set of 3D vectors for axial directions, and translate
-void Matrix4x4_FromVectors(mat4_t *out, const float vx[3], const float vy[3], const float vz[3], const float t[3]);
-
-// transforms a 3D vector through a matrix4x4
-void Matrix4x4_Transform (const mat4_t *in, const float v[3], float out[3]);
-// transforms a positive distance plane (A*x+B*y+C*z-D=0) through a rotation or translation matrix
-void Matrix4x4_TransformPositivePlane (const mat4_t *in, float x, float y, float z, float d, float *o);
-// transforms a Quake plane through a rotation or translation matrix
-void Matrix4x4_TransformQuakePlane(const mat4_t *in, const vec3_t n, float d, vec4_t *out);
-
-// ease of use functions
-// immediately applies a Translate to the matrix
-void Matrix4x4_ConcatTranslate (mat4_t *out, double x, double y, double z);
-// immediately applies a Rotate to the matrix
-void Matrix4x4_ConcatRotate (mat4_t *out, double angle, double x, double y, double z);
-// immediately applies a Scale to the matrix
-void Matrix4x4_ConcatScale (mat4_t *out, double x);
-// immediately applies a Scale3 to the matrix
-void Matrix4x4_ConcatScale3 (mat4_t *out, double x, double y, double z);
-
-// extracts scaling factor from matrix (only works for uniform scaling)
-double Matrix4x4_ScaleFromMatrix (const mat4_t *in);
-
-// scales vectors of a matrix in place and allows you to scale origin as well
-void Matrix4x4_Scale (mat4_t *out, double rotatescale, double originscale);
-
-// generate a perspective transformation matrix from bounds
-void Matrix4x4_FromFrustum (mat4_t *out, double left, double right, double bottom, double top, double nearval,
-                            double farval);
-// generate an orthogonal projection matrix from bounds
-void Matrix4x4_FromOrtho (mat4_t *out, double left, double right, double bottom, double top, double nearval,
-                          double farval);
+/**
+ * @return A `mat4_t` with the specified components.
+*/
+static inline mat4_t __attribute__ ((warn_unused_result)) Mat4(const float elements[16]) {
```

Comment:
The rest of the code puts __attribute__ after the function parameters (at the end of the line.). Let's pick one place for them and make it consistent.

Reply:
They can't be put at the end if it's a definition, only a declaration, that's why I moved them to the start.

Reply:
Oh! Weird!